_automation/main.go

```diff
@@ -0,0 +1,534 @@
+
+func init() {
+	defaultLogger = NewLogger()
+}
+
+const grammarsJson = "./_automation/grammars.json"
+
+type GrammarVersion struct {
+	Reference string `json:"reference"`
+	Revision  string `json:"revision"`
+}
+
+type Grammar struct {
+	Language string   `json:"language"`
+	URL      string   `json:"url"`
+	Files    []string `json:"files"`
+	*GrammarVersion
+}
+
+func (g *Grammar) ContentURL() string {
+	return fmt.Sprintf(
+		"https://raw.githubusercontent.com/%s",
+		strings.TrimPrefix(g.URL, "https://github.com/"),
+	)
+}
+
+func (g *Grammar) FetchNewVersion() *GrammarVersion {
+	if strings.HasPrefix(g.Reference, "v") {
+		tag, rev := fetchLastTag(g.URL)
+		if tag != g.Reference {
+			return &GrammarVersion{
+				Reference: tag,
+				Revision:  rev,
+			}
+		}
+	} else {
+		rev := fetchLastCommit(g.URL, g.Reference)
+		if rev != g.Revision {
+			return &GrammarVersion{
+				Reference: g.Reference,
+				Revision:  rev,
+			}
+		}
+	}
+	return nil
+}
+
+func root(args []string) error {
+	if len(args) < 1 {
+		return fmt.Errorf("sub-command expected")
+	}
+
+	ctx := context.Background()
+	s := NewUpdateService()
+
+	switch args[0] {
+	case "check-updates":
+		fs := flag.NewFlagSet("check-updates", flag.ExitOnError)
+		flagsParse(fs, args[1:])
+
+		s.CheckUpdates(ctx)
+	case "update":
+		if len(args) < 2 {
+			return fmt.Errorf("language argument is missing")
+		}
+
+		fs := flag.NewFlagSet("update", flag.ExitOnError)
+		force := fs.Bool("force", false, "re-download grammar even if revision is the same")
+		flagsParse(fs, args[2:])
+
+		s.Update(ctx, args[1], *force)
+	case "update-all":
+		fs := flag.NewFlagSet("update-all", flag.ExitOnError)
+		flagsParse(fs, args[1:])
+
+		s.UpdateAll(ctx)
+	default:
+		return fmt.Errorf("unknown sub-command")
+	}
+
+	return nil
+}
+
+func main() {
+	if err := root(os.Args[1:]); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+}
+
+type UpdateService struct {
+	grammars    []*Grammar
+	grammarsMap map[string]*Grammar
+}
+
+func NewUpdateService() *UpdateService {
+	b, err := os.ReadFile(grammarsJson)
+	if err != nil {
+		logAndExit(defaultLogger, err.Error(), "file", grammarsJson)
+	}
+
+	var grammars []*Grammar
+	err = json.Unmarshal(b, &grammars)
+	if err != nil {
+		logAndExit(defaultLogger, fmt.Sprintf("unmarshaling grammars file error: %w", err), "file", grammarsJson)
+	}
+
+	grammarsMap := make(map[string]*Grammar, len(grammars))
+	for _, g := range grammars {
+		grammarsMap[g.Language] = g
+	}
+
+	return &UpdateService{grammars, grammarsMap}
+}
+
+func (s *UpdateService) CheckUpdates(ctx context.Context) {
+	newVersions := s.fetchNewVersions()
+	for i, g := range s.grammars {
+		nextVersion := newVersions[i]
+		remoteReference := g.Reference
+		outdatedStr := ""
+		if nextVersion != nil {
+			remoteReference = nextVersion.Reference
+			outdatedStr = "outdated"
+		}
+		fmt.Printf("%s\t\tvendored: %s\tremote: %s\t%s\n", g.Language, g.Reference, remoteReference, outdatedStr)
+	}
+}
```

Comment:
Using a tabwriter would make the alignment of the columns better:

```suggestion
func (s *UpdateService) CheckUpdates(ctx context.Context) {
	newVersions := s.fetchNewVersions()
	w := tabwriter.NewWriter(os.Stdout, 0, 0, 3, ' ', 0)
	for i, g := range s.grammars {
		nextVersion := newVersions[i]
		fmt.Fprintf(w, "%s\t%s\t", g.Language, g.Reference)
		if nextVersion == nil {
			fmt.Fprintf(w, "(up-to-date)")
		} else if g.Reference == nextVersion.Reference {
			fmt.Fprintf(w, "(update available: %s -> %s)", g.Revision, nextVersion.Revision)
		} else {
			fmt.Fprintf(w, "(update available: %s -> %s)", g.Reference, nextVersion.Reference)
		}
		fmt.Fprintln(w)
	}
	w.Flush()
}
```

Reply:
This would be the output then:

```
bash         v0.19.0   (up-to-date)
c            v0.20.2   (up-to-date)
csharp       v0.20.0   (up-to-date)
cpp          v0.20.0   (up-to-date)
css          v0.19.0   (up-to-date)
cue          main      (up-to-date)
dockerfile   v0.1.2    (up-to-date)
elixir       main      (update available: 05e3631c6a0701c1fa518b0fee7be95a2ceef5e2 -> 869dff3ceb8823ca4b17ca33b663667c8e41e8ba)
elm          v5.3.7    (update available: v5.3.7 -> v5.6.3)
golang       master    (update available: 05900faa3cdb5d2d8c8bd5e77ee698487e0a8611 -> 64457ea6b73ef5422ed1687178d4545c3e91334a)
hcl          main      (update available: a8bd2568d37065a3b451cfd91745607fac4aac84 -> bdc2ee0cbd19d22f8a1529e5c9d099029573c931)
html         v0.19.0   (up-to-date)
java         v0.20.1   (up-to-date)
javascript   v0.19.0   (up-to-date)
kotlin       main      (update available: cce9073807c6c1da81ce658d683fdac2878e5b57 -> 826ef28d605d0925a86a99022cd222c96f2d0952)
lua          master    (update available: 9ec837de6061c05d4f202047e3e9dcf644c779e8 -> 0e860f697901c9b610104eb61d3812755d5211fc)
ocaml        v0.19.0   (update available: v0.19.0 -> v0.20.1)
php          v0.19.0   (up-to-date)
protobuf     main      (up-to-date)
python       v0.20.0   (up-to-date)
ruby         v0.19.0   (up-to-date)
rust         v0.20.1   (update available: v0.20.1 -> v0.20.3)
scala        v0.19.0   (update available: v0.19.0 -> v0.19.1)
svelte       v0.9.0    (update available: v0.9.0 -> v0.10.1)
toml         v0.5.1    (up-to-date)
typescript   v0.20.1   (update available: v0.20.1 -> v0.20.2)
yaml         v0.5.0    (up-to-date)
```

Reply:
Thank you